// @sap-ux/preview-middleware/dist/client/adp/quick-actions/fe-v4/create-table-custom-column.js
"use strict";

sap.ui.define(["sap/ui/dt/OverlayRegistry", "../../init-dialogs", "../table-quick-action-base", "../fe-v2/create-table-custom-column"], function (OverlayRegistry, ____init_dialogs, ___table_quick_action_base, ___fe_v2_create_table_custom_column) {
  "use strict";

  const DialogNames = ____init_dialogs["DialogNames"];
  const handler = ____init_dialogs["handler"];
  const SMART_TABLE_TYPE = ___table_quick_action_base["SMART_TABLE_TYPE"];
  const GRID_TABLE_TYPE = ___table_quick_action_base["GRID_TABLE_TYPE"];
  const MDC_TABLE_TYPE = ___table_quick_action_base["MDC_TABLE_TYPE"];
  const TableQuickActionDefinitionBase = ___table_quick_action_base["TableQuickActionDefinitionBase"];
  const TREE_TABLE_TYPE = ___table_quick_action_base["TREE_TABLE_TYPE"];
  const preprocessActionExecution = ___fe_v2_create_table_custom_column["preprocessActionExecution"];
  const CREATE_TABLE_CUSTOM_COLUMN = 'create-table-custom-column';
  const CONTROL_TYPES = [SMART_TABLE_TYPE, MDC_TABLE_TYPE, TREE_TABLE_TYPE, GRID_TABLE_TYPE];
  class AddTableCustomColumnQuickAction extends TableQuickActionDefinitionBase {
    constructor(context) {
      super(CREATE_TABLE_CUSTOM_COLUMN, CONTROL_TYPES, 'QUICK_ACTION_ADD_CUSTOM_TABLE_COLUMN', context);
    }
    async execute(path) {
      const {
        table,
        iconTabBarFilterKey,
        sectionInfo
      } = this.tableMap[path];
      if (!table) {
        return [];
      }
      preprocessActionExecution(table, sectionInfo, this.iconTabBar, iconTabBarFilterKey);
      this.selectOverlay(table);
      const overlay = OverlayRegistry.getOverlay(table);
      await handler(overlay, this.context.rta, DialogNames.ADD_FRAGMENT, undefined, {
        aggregation: 'columns',
        title: 'QUICK_ACTION_ADD_CUSTOM_TABLE_COLUMN'
      });
      return [];
    }
  }
  var __exports = {
    __esModule: true
  };
  __exports.CREATE_TABLE_CUSTOM_COLUMN = CREATE_TABLE_CUSTOM_COLUMN;
  __exports.CONTROL_TYPES = CONTROL_TYPES;
  __exports.AddTableCustomColumnQuickAction = AddTableCustomColumnQuickAction;
  return __exports;
});
