// @sap-ux/preview-middleware/dist/client/adp/quick-actions/fe-v4/table-quick-action-base.js
'use strict';
sap.ui.define([
    'sap/ui/fl/apply/api/FlexRuntimeInfoAPI',
    'open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common',
    '../../../cpe/quick-actions/utils'
], function (FlexRuntimeInfoAPI, ___sap_ux_private_control_property_editor_common, _____cpe_quick_actions_utils) {
    'use strict';
    const NESTED_QUICK_ACTION_KIND = ___sap_ux_private_control_property_editor_common['NESTED_QUICK_ACTION_KIND'];
    const getRelevantControlFromActivePage = _____cpe_quick_actions_utils['getRelevantControlFromActivePage'];
    const ACTION_ID = 'CTX_SETTINGS0';
    const CONTROL_TYPE = 'sap.ui.mdc.Table';
    class TableQuickActionDefinitionBase {
        kind = NESTED_QUICK_ACTION_KIND;
        get id() {
            return `${ this.context.key }-${ this.type }`;
        }
        get textKey() {
            return this.defaultTextKey;
        }
        isApplicable = false;
        isDisabled = false;
        get tooltip() {
            return undefined;
        }
        isClearButtonEnabled = false;
        children = [];
        tableMap = {};
        constructor(type, defaultTextKey, context, isSkipVariantManagementCheck) {
            this.type = type;
            this.defaultTextKey = defaultTextKey;
            this.context = context;
            this.isSkipVariantManagementCheck = isSkipVariantManagementCheck;
        }
        async initialize() {
            let index = 0;
            for (const smartTable of getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, [CONTROL_TYPE])) {
                if (!this.isSkipVariantManagementCheck) {
                    const hasVariantManagement = FlexRuntimeInfoAPI.hasVariantManagement({ element: smartTable });
                    if (!hasVariantManagement) {
                        continue;
                    }
                }
                const actions = await this.context.actionService.get(smartTable.getId());
                const changeColumnAction = actions.find(action => action.id === ACTION_ID);
                if (changeColumnAction) {
                    this.children.push({
                        label: `'${ smartTable.getHeader() }' table`,
                        enabled: true,
                        children: []
                    });
                    this.tableMap[`${ this.children.length - 1 }`] = index;
                    index++;
                }
            }
            if (this.children.length > 0) {
                this.isApplicable = true;
            }
        }
        getActionObject() {
            return {
                kind: NESTED_QUICK_ACTION_KIND,
                id: this.id,
                enabled: !this.isDisabled,
                tooltip: this.tooltip,
                title: this.context.resourceBundle.getText(this.textKey),
                children: this.children
            };
        }
    }
    var __exports = { __esModule: true };
    __exports.TableQuickActionDefinitionBase = TableQuickActionDefinitionBase;
    return __exports;
});
