// @sap-ux/preview-middleware/dist/client/adp/quick-actions/common/create-page-action.js
"use strict";

sap.ui.define(["sap/ui/dt/OverlayRegistry", "../../init-dialogs", "../simple-quick-action-base", "../../../utils/application", "../../../utils/version"], function (OverlayRegistry, ____init_dialogs, ___simple_quick_action_base, _____utils_application, _____utils_version) {
  "use strict";

  const DialogNames = ____init_dialogs["DialogNames"];
  const handler = ____init_dialogs["handler"];
  const SimpleQuickActionDefinitionBase = ___simple_quick_action_base["SimpleQuickActionDefinitionBase"];
  const getApplicationType = _____utils_application["getApplicationType"];
  const getUi5Version = _____utils_version["getUi5Version"];
  const isLowerThanMinimalUi5Version = _____utils_version["isLowerThanMinimalUi5Version"];
  const ADD_PAGE_ACTION = 'add-page-action';
  const CONTROL_TYPES = ['sap.f.DynamicPageTitle', 'sap.uxap.ObjectPageHeader', 'sap.uxap.ObjectPageDynamicHeaderTitle'];

  /**
   * Quick Action for adding a custom page action.
   */
  class AddPageActionQuickAction extends SimpleQuickActionDefinitionBase {
    constructor(context) {
      super(ADD_PAGE_ACTION, CONTROL_TYPES, 'QUICK_ACTION_ADD_CUSTOM_PAGE_ACTION', context);
    }
    async initialize() {
      const appType = getApplicationType(this.context.rta.getRootControlInstance().getManifest());
      const version = await getUi5Version();
      if (appType === 'fe-v4' && isLowerThanMinimalUi5Version(version, {
        major: 1,
        minor: 130
      })) {
        return;
      }
      return super.initialize();
    }
    async execute() {
      if (this.control) {
        const overlay = OverlayRegistry.getOverlay(this.control) || [];
        await handler(overlay, this.context.rta, DialogNames.ADD_FRAGMENT, undefined, {
          aggregation: 'actions',
          title: 'QUICK_ACTION_ADD_CUSTOM_PAGE_ACTION'
        });
      }
      return [];
    }
  }
  var __exports = {
    __esModule: true
  };
  __exports.ADD_PAGE_ACTION = ADD_PAGE_ACTION;
  __exports.AddPageActionQuickAction = AddPageActionQuickAction;
  return __exports;
});
